### ISSUE TITLE
tf.random.uniform generates identical random numbers when used  both inside and outside tf.while_loop  in a decorated function

### ISSUE THREAD

**Issue by vipeano** [tensorflow/tensorflow#58587]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

2.8.1

### Custom Code

No

### OS Platform and Distribution

Mac Os Monterey

### Current Behaviour?

```shell
When tf.random.uniform is called both inside and outside a tf.while_loop in a function decorated with tf.function, its internal counter is the same for both calls generating the same number. The expected behavior (which occur without decoration) is the internal counter to change and, thus, tf.random.uniform to generate two different outputs.
```

### Standalone code to reproduce the issue

```shell
max_steps=1

def condition(index,  random_num):
    return index < max_steps
def body(index,  random_num):
    return index+1, tf.random.uniform([1],dtype='float32')
@tf.function   
def function():
    index = tf.constant(0, dtype="int8")
    
    a=tf.random.uniform([1],dtype='float32')    
  
    index, random_num=tf.while_loop(condition,body,[index, tf.zeros([1],dtype='float32')])
    
#    b = tf.random.uniform([1])
#    c = tf.random.uniform([1])
    return a, random_num

a, b=function()

print(a, b) 

# the same random number is assigned both to  a and b (without decoration a and b would be two different random numbers)
```

### Relevant log output

_No response_</details>

**Comment by vipeano** [comment]
A workaround to this bug is not to use tf.random.uniform() as in the code below
```
max_steps=1
g = tf.random.Generator.from_non_deterministic_state()

def condition(index,  random_num):
    return index < max_steps
def body(index,  random_num):
    
    return index+1, g.uniform([1])

@tf.function   
def function():

    index = tf.constant(0, dtype="int8")
    
    a=g.uniform([1])    
  
    index, random_num=tf.while_loop(condition,body,[index, tf.zeros([1],dtype='float32')])
    

    return a, random_num

a, b=function()

print(a, b) 
```

**Comment by bhack** [comment]
@vipeano Can you reproduce the same with the last stable TF 2.10 version?

**Comment by SuryanarayanaY** [comment]
Hi @vipeano ,

I tried to replicate the issue you mentioned with the provided code.But Iam getting different output for both `a & b` with 2.8.1V and 2.9.2V also which conflicts your claim.I even tried without @tf.function decorator even then the output is different for both `a & b`.Please refer to attached gist [here](https://colab.research.google.com/gist/SuryanarayanaY/ed203f402daf6cd7e5faa67d5406b571/58587.ipynb).

If you still able to replicate the mentioned behaviour please submit a gist for further investigation.

Thankyou!

**Comment by vipeano** [comment]
Hi,

thank you to pointing to this bug in my bug report :-)

The code I posted was incomplete. The bug only appears  if one is also fixing the global seed with tf.keras.utils.set_random_seed() 

[Here](https://colab.research.google.com/drive/1apyhAobce0RAGpOThyu_RAFn7peXJhzU#scrollTo=iF6F6pSZMO2d) is a link to a colab reproducing the bug (it is there also in tf. 2.9.2).

Best,

Vittorio

**Comment by SuryanarayanaY** [comment]
Hi @vipeano ,

TF has two random seeds.One at `global level` as `tf.random.set_seed(seed)` and other at `operation level` like `tf.random.uniform(...,seed=xyz)`.Both these together controls the random seeding. `tf.while_loop` has a default `keyword argument ` `parallel_iterations=10` and combined with `tf.function` decorator is applied which disables Eager execution inside the function, it is maintaining same state till `10 iterations` which is the expected behaviour. If we set the `max_steps=11` then we can see the change in the output.We can also decrease the `parallel_iterations=1 and max_steps=2` to observe the change in behaviour.

Please refer the attached [gist](https://colab.research.google.com/gist/SuryanarayanaY/00a09dd6fa5107e8ab3ddbd9f8e23074/58587.ipynb) for checking the behaviour w.r.t `parallel_iterations`.

Please refer the sources [while_loop](https://www.tensorflow.org/api_docs/python/tf/while_loop?version=nightly), and [set_seed](https://www.tensorflow.org/api_docs/python/tf/random/set_seed) for more details.

Thankyou.
